src/topology/ping_pong.rs

```diff
@@ -0,0 +1,865 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Implements the Ping-Pong Topology described in [VDAF]. This topology assumes there are exactly
+//! two aggregators, designated "Leader" and "Helper". Note that while this implementation is
+//! compatible with VDAF-06, it actually implements the ping-pong wrappers specified in VDAF-07
+//! (forthcoming) since those line up better with the VDAF implementation in this crate.
+//!
+//! [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+
+use crate::{
+    codec::{decode_u32_items, encode_u32_items, CodecError, Decode, Encode, ParameterizedDecode},
+    vdaf::{Aggregator, PrepareTransition, VdafError},
+};
+use std::fmt::Debug;
+
+/// Errors emitted by this module.
+#[derive(Debug, thiserror::Error)]
+pub enum PingPongError {
+    /// Error running prepare_init
+    #[error("vdaf.prepare_init: {0}")]
+    VdafPrepareInit(VdafError),
+
+    /// Error running prepare_preprocess
+    #[error("vdaf.prepare_preprocess {0}")]
+    VdafPreparePreprocess(VdafError),
+
+    /// Error running prepare_step
+    #[error("vdaf.prepare_step {0}")]
+    VdafPrepareStep(VdafError),
+
+    /// Error decoding a prepare share
+    #[error("decode prep share {0}")]
+    CodecPrepShare(CodecError),
+
+    /// Error decoding a prepare message
+    #[error("decode prep message {0}")]
+    CodecPrepMessage(CodecError),
+
+    /// State machine mismatch between peer and host.
+    #[error("state mismatch: host state {0} peer state {0}")]
+    StateMismatch(&'static str, &'static str),
+
+    /// Internal error
+    #[error("internal error: {0}")]
+    InternalError(&'static str),
+}
+
+/// Distinguished aggregator roles.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Role {
+    /// The Leader aggregator.
+    Leader,
+    /// The Helper aggregator.
+    Helper,
+}
+
+/// Corresponds to `struct Message` in [VDAF's Ping-Pong Topology][VDAF]. All of the fields of the
+/// variants are opaque byte buffers. This is because the ping-pong routines take responsibility for
+/// decoding preparation shares and messages, which usually requires having the preparation state.
+///
+/// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+#[derive(Clone, PartialEq, Eq)]
+pub enum Message {
```

Comment:
Elsewhere we use distinctive names that reflect the module in which the type is defined.
```suggestion
pub enum PingPongMessage {
```